bluesky/plans.py

```diff
@@ -36,6 +36,17 @@ def count(detectors, num=1, delay=None, *, md=None):
         If None, capture data until canceled
     delay : iterable or scalar, optional
         Time delay in seconds between successive readings; default is 0.
+    per_step : callable, optional
```

Comment:
I don't think this signature should expect `step` and `pos_cache`. Notice the signatures of the existing `per_step` callables are suited to their purposes:

```
one_1d_step(detectors, motor, step)
one_nd_step(detectors, step, pos_cache)
```

I think this should default to plain `trigger_and_read` if `None`, with no need for a special `count_step` function necessary, and accept anything that has the same signature as `trigger_and_read`.

Reply:
@danielballan the reason I have included `step` and `pos_cache` as arguments is so that I can use the _same_  `per_step` function in `scan` or `count` as outlined in the motivation and context above. This is fundamental to the current issue that this PR is looking to resolve.

FYI. I think all of the built in plans, with the exception of `list_scan`, use `one_nd_step` as the default 'per_step' function therefore I would propose that we modify `one_1d_step` to use the same signature as this and `one_nd_step`.

Reply:
In fact after looking in more detail, one_1d_step appears to be a specific case of one_nd_step so I think we could deprecate one_1d_step entirely.

Reply:
. Yes, once https://github.com/NSLS-II/bluesky/issues/1144 is done we can deprecate `one_1d_step`. I can see your point that having one `per_step` interface is advantageous and generally in keeping with the spirit of limiting the number of interfaces. I wonder a little if, as a user of `count`, I would wonder why my `per_step` function has to accept two parameters that it doesn't actually use. Perhaps we should use signature inspection to accept either `one_nd_step`-like functions or `trigger_and_read`-like functions. See `scan` for an example of this. Thoughts?

Reply:
are you suggesting that we use the system we use in `scan` (and `grid_scan`) in order to accept an arbitrary number of `motor, start, stop` arguments? 

If so I am not sure that is the best path, that would mean that every `per_step` function would need complicated logic in order to deal with various numbers of parameters being passed in. The solution I am proposing here is to say that the `plans` know what parameters are useful and provide empty dictionaries for those that are not useful. This seems like the most simple solution, and because the 'extra' parameters both relate to moving motors passing in empty dictionaries is equivalent to saying 'move nothing'.

In fact passing empty `dicts` for `step` and `scan` in `count` to `one_nd_scan` would also work, so that would also negate the need for a custom default `per_step` for count.

Reply:
Not exactly. I'm suggesting that the `per_step` in `count` specifically could accept the standard `one_nd_step` interface or the standard `trigger_and_read` interface. Everything else would require the `one_nd_step` interface (and also, just for back-compat reasons, `one_1d_step`).

I'm not sure I like that better than what you have proposed here -- just brainstorming what our options are. Any opinions, @ZLLentz @teddyrendahl?

Reply:
ok so now I think I understand, you are suggesting having some logic in `count` that works out what parameters to send to the `per_step` function based on what the function is. is that it?

I still think it is cleaner and neater to pass in the emtpy `dicts`, but as always will follow the consensus.

Reply:
Yes, that is he possibility I want to consider. I’m not sure I disagree with your view to have only one “per_step” signature, just want to consider it.

Reply:
This is making the assumption that all per-step will be happy with empty dictionaries as input.

I think that it is clearer here if `count` has it's own `per_step` signature.  If you can re-use another `per_step` function, it is easy enough to use `partial` to project out the extra inputs that `count` does not provide (or make them have default values).

If we set the precedent that all `per_step` must have the same signature, what do we do if we come up with a plan that requires a _more_ complex signature?

On the other hand, a `per_step` that is only meant for `count` (ex, https://github.com/NSLS-II/bluesky/pull/1149/files#diff-ef3b7c7e4a0df009f8b37ec3d866bfbbR895) that will happily go into the `one_1d_step` slot but very much not work.   It is possible to defensively write your count-only per_step to raise if the second two arguments are not empty, but it seems clearer to let cpython raise a `TypeError` and clearly explain what the problem is.

Reply:
> I think that it is clearer here if count has it's own per_step signature. If you can re-use another per_step function, it is easy enough to use partial to project out the extra inputs that count does not provide (or make them have default values).

I think the exact opposite is true, why introduce complexity (.i.e. a user has to know that for plan 'a' you need this signature but for plan 'b' this different signature, or for plan 'c' ........). Which I think complicates things to much.

Aside from user confusion when we have to build these plans in factories from code (i.e. behind the scenes of a potential GUI) then this also becomes increasingly complicated as you have to consider each case individually.....

> If we set the precedent that all per_step must have the same signature, what do we do if we come up with a plan that requires a more complex signature?

This is a fair point, although I guess we could add extra info via `kwargs` (i.e. add `**kwargs` to the signature) which would be selectively ignored by some `per_step` functions but not `others`.

> On the other hand, a per_step that is only meant for count (ex, https://github.com/NSLS-II/bluesky/pull/1149/files#diff-ef3b7c7e4a0df009f8b37ec3d866bfbbR895) that will happily go into the one_1d_step slot but very much not work.

@danielballan and I had this conversation above, I think we don't need `count_step` (or `one_1D_step` for that matter anymore as `one_nd_step` will work in all our use cases as is (it would require some changes to the plans that currently use `one_1D_step` but that is already planned as mentioned by @danielballan above).
If someone insists on writing a `count_only` `per_step` plan then I think the onus is on them to ensure that it raises the correct error, especially when making it universal (in terms of the plans we currently have) is as simple as adding the one line `yield from move_per_step(step, pos_cache)` as is found here https://github.com/NSLS-II/bluesky/blob/226d74f97ae34963733bf725b2eed5d2790ffe88/bluesky/plan_stubs.py#L951

Reply:
> If we set the precedent that all per_step must have the same signature, what do we do if we come up with a plan that requires a more complex signature?

I think this is convincing.

> I think we don't need count_step ... one_nd_step will work in all our use cases as is

I think this assumes too much about what other types of plans might be added in the future, unforeseen at this point.

> we could add extra info via kwargs... which would be selectively ignored by some per_step functions but not others

This pattern is one we avoid at all costs. It falls under the category of "silently ignoring user input" which can lead to very confusing traps.

> I think it [would be less clear to give `count`'s `per_step` its own signature], why introduce complexity (.i.e. a user has to know that for plan 'a' you need this signature but for plan 'b' this different signature, or for plan 'c' ........). Which I think complicates things to much.

I can see how the "complication" argument can cut both ways. Looking at the library as a whole, "one per_step interface" has a simplicity to it, as you argue. But approach it as a user of `count` trying to write my own `per_step` funciton, I would wonder why my function needs to accept `step` and `pos_cache` ("What even are those?") only to ignore them. (Apologies for repeating myself here.)

In summary given that
) complication can cut both ways
* silently ignoring user inputs is known to create deeply confusing "bugs"
( we can't be sure that the `one_nd_step` signature is general enough to cover all future plans

I think there is a strong case for making `count`'s `per_step` expect the signature of `trigger_and_read`.

I am still interested in input from @ZLLentz and @teddyrendahl if you guys have time to weigh in.

Reply:
I reiterate my stance above that I think having different signatures for `per_step` is _more_ confusing than having some which_ignore_ some parameters. I am completely not convinced of the arguments above. 

This suggestion is moving a small level of complexity, in ensuring that the `per_step` functions all have the same signature, to ensuring that both users and code which wants to work with all plans and a `per_step` function have to generate a multitude of `per_step` functions and know which plan to send them too.

Reply:
Consider the logical extension of that argument. We can't just have one `bluesky.plans.plan` that takes every possible argument and ignores ones that are not applicable.

Reply:
It might be worth spelling out the alternative @tacaswell and I are now advocating, just to make sure we are all on the same page.

There would be two `per_step` signatures in bluesky (plus the one deprecated one, supported for back-compat) with room to grow more. I can't picture us growing a _multitude_ of more signatures, because there aren't that many fundamentally different kinds of plan likely to be added to bluesky, but I could picture a couple more in a long term.

Thus, the use cases and associated user experiences would be:

1. I need to customize `count`. I check the docstring. It says, "`per_step` accepts a plan with the signature `f(detector)`. The default is `bluesky.plan_stubs.trigger_and_read`." I do a customization like:

    ```py
    def my_trigger_and_read(detectors):
        yield from bluesky.plan_stubs.trigger_and_read(detector)
        print('hi')
    ```

2. I need to customize a 1D or ND scan.  I check the docstring. It says, "`per_step` accepts a plan with the signature `f(detectors, step, pos_cache)`. The default is `bluesky.plan_stubs.one_nd_step`." I I do a customization like:
    ```py
    def my_one_nd_step(detectors, step, pos_cache):
        yield from bluesky.plan_stubs.one_nd_step(detectors, step, pos_cache)
        print(f'finished step {step}')

3. But what I want to use `my_one_nd_step` on `count`? The thing is, this customization makes explicit use of `step` so I _can't_ just plug it into `count` as is. It does something that doesn't make sense to do for `count` because `count` doesn't scan movable things in steps.

4. OK, what if I want to use `my_trigger_and_read` in a scan? That one makes more sense: you have to copy `one_nd_step` and replace the `trigger_and_read` call in its inner loop for `my_trigger_and_read`.

In summary, making one signature to rule them all doesn't make the user experience simpler; it just buries the real difference where they are harder to see.

Reply:
Some rambling thoughts, note that I don't have strong opinions about this:
- I agree with Dan that it isn't necessary to require extra parameters that are effectively always receiving empty dictionaries. The important thing is that the parameters have the same names when they are relevant.
- I think the problem with not being able to plug `my_one_nd_step` into `count` is highlighting a problem I have with `scan_nd`'s `per_step` parameter: in general what we really want is just to modify or append to the `trigger_and_read` step. Any `per_step` I've written simply passes the `step` and `pos_cache` args directly into `move_per_step`, so if there weren't compelling use cases for these args I'd actually prefer to have `per_step(detectors)` be the norm everywhere. By using `scan_nd` I'm implying that I want to do the `move_per_step` step scan. But there are compelling use-cases here, so it's fine.
- In general, I don't like `per_step` options at all because they make it very easy to make plans with misleading metadata. We'll have plans named `count` that aren't really counting... Plans named `scan_nd` that aren't really scanning. etc.

Reply:
@teddyrendahl the reason for passing in the the positions is if you need to implement a scan that has complex "these can move in parallel, these must move in series" constraints.

Reply:
@tacaswell at the risk of bringing this comment chain too far off-topic, I see the use cases but still find the API of "`scan_nd` decides what points we use based on my args, and then I also pass a `per_step` to re-interpret them" to be less intuitive than "I write a new scan that decides what complex set of points I use"

Back on-topic: I agree that count's `per_step` should just have the `detectors` argument. We should think of `scan_nd` as "adding" arguments to `per_step` rather than thinking of `count` as "removing" arguments from `per_step`.